functional-samples/ai.gemini-on-device-calendar-mate/background.js

```diff
@@ -0,0 +1,146 @@
+/* global LanguageModel */
+
+import anyDateParser from './node_modules/any-date-parser/dist/index.mjs';
+
+chrome.runtime.onInstalled.addListener(() => {
+  chrome.contextMenus.create({
+    id: 'createCalendarEvent',
+    title: 'Create Calendar Event',
+    contexts: ['selection']
+  });
+});
+
+chrome.contextMenus.onClicked.addListener(async (info) => {
+  if (info.menuItemId === 'createCalendarEvent') {
+    const selectedText = info.selectionText;
+    try {
+      // Extract event details from text using the prompt API
+      const event = await parseEventDetails(selectedText);
+      // Convert date to RFC 5545 date format
+      const startDate = parse(
+        [
+          event.start_time,
+          event.start_date,
+          event.start_year,
+          event.timezone
+        ].join(' ')
+      );
+      const endDate = parse(
+        [event.end_time, event.end_date, event.end_year, event.timezone].join(
+          ' '
+        )
+      );
+      event.dates = format(startDate) + '/' + format(endDate);
+
+      const googleCalendarUrl = createGoogleCalendarUrl(event);
+      chrome.tabs.create({ url: googleCalendarUrl });
+    } catch (e) {
+      console.log(e);
+    }
+  }
+});
+
+function parse(dateString) {
+  const parsed = anyDateParser.attempt(dateString);
+  return new Date(
+    parsed.year,
+    parsed.month - 1,
+    parsed.day,
+    parsed.hour,
+    parsed.minute
+  );
+}
+
+function format(date) {
+  let month = '' + (date.getMonth() + 1);
+  let day = '' + date.getDate();
+  const year = date.getFullYear();
+
+  if (month.length < 2) month = '0' + month;
+  if (day.length < 2) day = '0' + day;
+  const hours = String(date.getHours()).padStart(2, '0');
+  const minutes = String(date.getMinutes()).padStart(2, '0');
+  const seconds = String(date.getSeconds()).padStart(2, '0');
+
+  return [year, month, day, 'T', hours, minutes, seconds].join('');
+}
+
+function createGoogleCalendarUrl(eventDetails) {
+  let googleCalendarUrl =
+    'https://calendar.google.com/calendar/render?action=TEMPLATE';
+
+  if (eventDetails.title) {
```

Comment:
We might be able to use [URLSearchParams](https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams) which would be more idiomatic.

Reply:
good call! done